pennylane/circuit_drawer/circuit_drawer.py

```diff
@@ -0,0 +1,372 @@
+        self.observable_decoration_indices = []
+
+        CircuitDrawer.move_multi_wire_gates(self.operation_grid)
+
+        # Resolve operator names
+        self.resolve_representation(self.operation_grid, self.operation_representation_grid)
+        self.resolve_representation(self.observable_grid, self.observable_representation_grid)
+
+        # Add multi-wire gate lines
+        self.operation_decoration_indices = self.resolve_decorations(
+            self.operation_grid, self.operation_representation_grid,
+        )
+        self.observable_decoration_indices = self.resolve_decorations(
+            self.observable_grid, self.observable_representation_grid,
+        )
+
+        CircuitDrawer.pad_representation(
+            self.operation_representation_grid,
+            charset.WIRE,
+            "",
+            2 * charset.WIRE,
+            self.operation_decoration_indices,
+        )
+
+        CircuitDrawer.pad_representation(
+            self.operation_representation_grid,
+            charset.WIRE,
+            "",
+            "",
+            set(range(self.operation_grid.num_layers)) - set(self.operation_decoration_indices),
+        )
+
+        CircuitDrawer.pad_representation(
+            self.observable_representation_grid,
+            " ",
+            charset.MEASUREMENT + " ",
+            " ",
+            self.observable_decoration_indices,
+        )
+
+        CircuitDrawer.pad_representation(
+            self.observable_representation_grid,
+            charset.WIRE,
+            "",
+            "",
+            set(range(self.observable_grid.num_layers)) - set(self.observable_decoration_indices),
+        )
+
+        self.full_representation_grid = self.operation_representation_grid.copy()
+        self.full_representation_grid.append_grid_by_layers(self.observable_representation_grid)
+
+    def make_wire_conversion_dicts(self, raw_operation_grid, raw_observable_grid):
+        """Prepare the dictionaries used to convert between internal and device wires.
+
+        This conversion is necessary due to the fact that the circuit drawer internally uses
+        ascending wires that have to be matched to the actual wires of the Operations inside
+        the circuit.
+
+        Args:
+            raw_operation_grid (Iterable[~.Operator]): The raw grid of operations
+            raw_observable_grid (Iterable[~.Operator]): The raw  grid of observables
+        """
+        # pylint: disable=protected-access
+        all_operators = list(qml.utils._flatten(raw_operation_grid)) + list(qml.utils._flatten(raw_observable_grid))
+        all_wires = [op.wires for op in all_operators if op is not None]
+        device_wires = sorted(set(qml.utils._flatten(all_wires)))
+        internal_wires = list(range(len(device_wires)))
+
+        self._device_wire_to_internal_wire = dict(zip(device_wires, internal_wires))
+        self._internal_wire_to_device_wire = dict(zip(internal_wires, device_wires))
```

Comment:
Could the ``inv_map = {v: k for k, v in my_map.items()}`` type of inverting a dictionary work here, once you have ``self._device_wire_to_internal_wire``? This mapping seems to be one-to-one, so it seems that there shouldn't be a problem with inversion.

Reply:
Of course one could do it like this, but as I have the keys and values at hand I prefer to do it this way around because it just reads easier.